Correct combining of masked images

jeremykubica opened the pull request:
The median function would set masked pixels to zero, which can skew the distribution. The new code just computes the median over the non-masked pixels.

Also adds a computeMeanImage function.

Adds corresponding tests.

jbkalmbach commented on search/src/RawImage.cpp:
```diff
@@ -432,16 +431,28 @@ RawImage createMedianImage(const std::vector<RawImage>& images)
+                                 pixArray.begin() + num_unmasked);
+                result.setPixel(x, y, pixArray[median_ind]);
+            } else {
+                result.setPixel(x, y, NO_DATA);
```
How does leaving pixels with `NO_DATA` in the stamps affect filtering? That's the only change here I see that affects things further down the pipeline.

jeremykubica replied:
That's a good point. The maximum value based filtering will work fine, but the CNN filtering might not work correctly anymore. I changed this to return a result of 0.0 in the case of all the pixels being no data to stay consistent with the old code.

jbkalmbach commented on tests/test_raw_image.py:
```diff
@@ -203,6 +246,50 @@ def test_create_summed_image(self):
+        self.assertAlmostEqual(mean_image.get_pixel(0, 2), 8.8/3.0, delta=1e-6)
+        self.assertAlmostEqual(mean_image.get_pixel(1, 2), 9.4/3.0, delta=1e-6)
+
+        # Apply masks to images 1 and 3.
```
Comment doesn't match code below.

jeremykubica replied:
Fixed.

jbkalmbach reviewed (approved):
Looks good. Thanks for the update, this change makes a lot of sense to add.

jeremykubica commented:
I fixed your comments and, in the process, realized that the changes needed to be propagated a bit further into the python code. I also added a bunch of tests and realized that the way we have been computing median is not consistent between the python and c++ code (if there are an even number of elements np.median takes the average of the values), so I corrected this as well.

Can you take another look when you get a chance?

jbkalmbach reviewed (approved):
New changes look good. In the future it would be good to think about the best way to handle any masked pixels downstream in our filtering pipeline and then revisit what we do here. Should they be 0? Or should we come up with an alternate value to tell whatever we are using for filtering that there is no data? Because a real value around 0 is actual information compared to a pixel with no data.
